Source/WebCore/style/ChildChangeInvalidation.cpp

```diff
@@ -112,28 +112,98 @@ void ChildChangeInvalidation::invalidateForHasBeforeMutation()
 {
     ASSERT(m_needsHasInvalidation);
 
-    if (m_childChange.isInsertion() && m_childChange.type != ContainerNode::ChildChange::Type::AllChildrenReplaced)
-        return;
-
     MatchingHasSelectors matchingHasSelectors;
 
     traverseRemovedElements([&](auto& changedElement) {
-        invalidateForChangedElement(changedElement, matchingHasSelectors);
+        invalidateForChangedElement(changedElement, matchingHasSelectors, ChangedElementRelation::SelfOrDescendant);
     });
+
+    auto firstChildStateWillStopMatching = [&](Element* elementAfterChange) {
```

Comment:
I think elementAfterChange / elementBeforeChange are a bit confusing (given before/after change can refer to the mutation), can we name them `elementNextSibling` / `elementPreviousSibling` or something?